modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusManager.scala

```diff
@@ -176,17 +179,36 @@ object ConsensusManager {
           .ifM(internalFacilitateWith(EventTrigger.some), Applicative[F].unit)
     }
 
-    Spawn[F]
-      .start(clusterStorage.peerChanges.evalMap {
-        case (peerId, Some(peer)) if peer.state === NodeState.Observing =>
-          val res = for {
-            exchangeRequest <- consensusStorage.getOwnRegistration.map(RegistrationExchangeRequest(_))
-            exchangeResponse <- consensusClient.exchangeRegistration(exchangeRequest).run(peer)
-            _ <- exchangeResponse.maybeKey.traverse(consensusStorage.registerPeer(peerId))
-          } yield ()
-          res.handleErrorWith(err => logger.error(err)(s"Error exchanging consensus registration with peer ${peerId.show}"))
-        case _ => Applicative[F].unit
-      }.compile.drain)
-      .as(manager)
+    Spawn[F].start(
+      nodeStorage.nodeStates
+        .filter(NodeState.leaving.contains)
+        .evalTap { _ =>
+          (consensusStorage.getLastKeyAndArtifact, consensusStorage.getOwnRegistration).mapN {
+            case (lastKey, ownRegistration) => lastKey.map(_._1).flatMap(k => ownRegistration.map((k, _)))
+          }.map {
+            _.map {
+              case (lastKey, ownRegistration) => Order[Key].max(lastKey.next, ownRegistration)
+            }
+          }
+            .map(_.map(Deregistration(_)))
+            .flatMap {
+              _.traverse(event => gossip.spread(event))
+            }
```

Comment:
Sorry for being annoying, but I think we can shorten this expression to:
```scala
(consensusStorage.getLastKey, consensusStorage.getOwnRegistration)
  .tupled
  .flatTap(_.mapN(Order[Key].max).map(Deregistration(_)).traverse(gossip.spread))
```

With the introduction of this function to `ConsensusStorage`:
```scala
def getLastKey: F[Option[Key]] = lastKeyAndArtifactR.get.map(_._1F)
```